src/test/java/org/prebid/server/bidder/loyal/LoyalBidderTest.java

```diff
@@ -0,0 +1,349 @@
+
+import java.util.List;
+import java.util.function.UnaryOperator;
+
+import static java.util.Arrays.asList;
+import static java.util.Collections.singletonList;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatIllegalArgumentException;
+import static org.prebid.server.proto.openrtb.ext.response.BidType.audio;
+import static org.prebid.server.proto.openrtb.ext.response.BidType.banner;
+import static org.prebid.server.proto.openrtb.ext.response.BidType.video;
+import static org.prebid.server.proto.openrtb.ext.response.BidType.xNative;
+
+public class LoyalBidderTest extends VertxTest {
+
+    private static final String ENDPOINT_URL = "https://test.com/test";
+
+    private final LoyalBidder target = new LoyalBidder(ENDPOINT_URL, jacksonMapper);
+
+    @Test
+    public void creationShouldFailOnInvalidEndpointUrl() {
+        assertThatIllegalArgumentException().isThrownBy(() -> new LoyalBidder("invalid_url", jacksonMapper));
+    }
+
+    @Test
+    public void makeHttpRequestsShouldMakeOneRequestPerImp() {
+        // given
+        final BidRequest bidRequest = BidRequest.builder()
+                .imp(asList(givenImp(UnaryOperator.identity()), givenImp(UnaryOperator.identity())))
+                .build();
+
+        //when
+        final Result<List<HttpRequest<BidRequest>>> result = target.makeHttpRequests(bidRequest);
+
+        //then
+        assertThat(result.getErrors()).isEmpty();
+        assertThat(result.getValue()).hasSize(2)
+                .extracting(HttpRequest::getPayload)
+                .extracting(BidRequest::getImp)
+                .extracting(List::size)
+                .containsOnly(1);
+        assertThat(result.getValue())
+                .extracting(HttpRequest::getPayload)
+                .flatExtracting(BidRequest::getImp)
+                .extracting(Imp::getId)
+                .containsExactlyInAnyOrder("123", "123");
+    }
+
+    @Test
+    public void shouldMakeOneRequestWhenOneImpIsValidAndAnotherIsNot() {
+        // given
+        final BidRequest bidRequest = BidRequest.builder()
+                .imp(asList(givenImp(UnaryOperator.identity()), givenBadImp(UnaryOperator.identity())))
+                .build();
+
+        //when
+        final Result<List<HttpRequest<BidRequest>>> result = target.makeHttpRequests(bidRequest);
+
+        //then
+        assertThat(result.getErrors().get(0).getMessage()).isEqualTo("Invalid imp: invalidImp");
+        assertThat(result.getValue()).hasSize(1)
+                .extracting(HttpRequest::getPayload)
+                .extracting(BidRequest::getImp)
+                .extracting(List::size)
+                .containsOnly(1);
+        assertThat(result.getValue())
+                .extracting(HttpRequest::getPayload)
+                .flatExtracting(BidRequest::getImp)
+                .extracting(Imp::getId)
+                .containsExactly("123");
```

Comment:
this check doesn't make any sense since all the imps have the same id

also, with the new check, you don't need the old one, because it's a duplicate

Reply:
Is there a example test like this in another bidder?

Reply:
I meant those two checks are duplicate

```
assertThat(result.getValue()).hasSize(1)
                .extracting(HttpRequest::getPayload)
                .extracting(BidRequest::getImp)
                .extracting(List::size)
                .containsOnly(1);

assertThat(result.getValue())
                .extracting(HttpRequest::getPayload)
                .flatExtracting(BidRequest::getImp)
                .extracting(Imp::getId)
                .containsExactly("123");
```